x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/cluster/routing/allocation/mapper/DataTierFieldMapper.java

```diff
@@ -73,7 +73,25 @@ public Query existsQuery(SearchExecutionContext context) {
 
         @Override
         public ValueFetcher valueFetcher(SearchExecutionContext context, String format) {
-            throw new UnsupportedOperationException("Cannot fetch values for internal field [" + name() + "].");
+            if (format != null) {
+                throw new IllegalArgumentException("Field [" + name() + "] of type [" + typeName() + "] doesn't support formats.");
+            }
+
+            String tierPreference = getTierPreference(context);
+            return lookup -> singletonList(tierPreference);
+        }
+
+        private String getTierPreference(SearchExecutionContext context) {
+            Settings settings = context.getIndexSettings().getSettings();
+            String value = DataTierAllocationDecider.INDEX_ROUTING_PREFER_SETTING.get(settings);
+
+            if (value == null) {
```

Comment:
I assume this will change after #71439 is in, so keeping this comment here for now just to recheck